src/Features/Core/Shared/Extensions/DiagnosticAnalyzerExtensions.cs

```diff
@@ -54,7 +54,12 @@ public static DiagnosticAnalyzerCategory GetDiagnosticAnalyzerCategory(this Diag
                         category |= DiagnosticAnalyzerCategory.ProjectAnalysis;
                     }
 
-                    if (!cantSupportSemanticSpanAnalysis)
+                    // Current DiagnosticAnalyzer API has no restriction on any analyzer action reporting diagnostics anywhere within a compilation.
+                    // Hence, we must conservatively assume they can't support incremental span based analysis across compilations.
+                    // GitHub issue https://github.com/dotnet/roslyn/issues/1565 tracks adding support for marking an analyzer incremental.
+                    // Until then, we just assume that all public/non-built in analyzers can never support incremental analysis.
+
+                    if ((analyzer is IBuiltInAnalyzer) && !cantSupportSemanticSpanAnalysis)
```

Comment:
okay, here is what I am proposing.

for third party analyzer, for solution crawler path (AnalyzeXXXX), don't treat any of them as supporting method body analysis. always do full document analysis.

for LB path (TryGetLatest ...), change code so that, if analyzer supports SyntaxNode action, Symbol action and etc, still let it be called and return diagnostics for the span but return "false" from the method which indicates it is not returning full diagnostics.

engine will take care of rest and call the diagnostic service again when it sees new diagnostics are found by solution crawler path.

this should make LB to show up fast for common case, but still not miss expensive diagnostic for LB.

if you change code like above, LB will never go cheap path for third party analyzer and only show LB when solution crawler path is done.